// electrode-electrify-react-component/lib/components/electrode-electrify-react-component.js
"use strict";

exports.__esModule = true;

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require("babel-runtime/helpers/possibleConstructorReturn");

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require("babel-runtime/helpers/inherits");

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _assets = require("./sub-components/assets");

var _assets2 = _interopRequireDefault(_assets);

var _modules = require("./sub-components/modules");

var _modules2 = _interopRequireDefault(_modules);

var _electrify = require("../../src/styles/electrify.css");

var _electrify2 = _interopRequireDefault(_electrify);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ElectrodeElectrifyReactComponent = function (_React$Component) {
  (0, _inherits3.default)(ElectrodeElectrifyReactComponent, _React$Component);

  function ElectrodeElectrifyReactComponent() {
    (0, _classCallCheck3.default)(this, ElectrodeElectrifyReactComponent);
    return (0, _possibleConstructorReturn3.default)(this, _React$Component.apply(this, arguments));
  }

  ElectrodeElectrifyReactComponent.prototype.render = function render() {
    return _react2.default.createElement(
      "div",
      null,
      _react2.default.createElement(
        "div",
        { className: _electrify2.default.leftColumn },
        _react2.default.createElement(_modules2.default, {
          modules: this.props.modules,
          modulesByPkg: this.props.modulesByPkg,
          totalSize: this.props.totalSizeByPkg
        })
      ),
      _react2.default.createElement(
        "div",
        { className: _electrify2.default.rightColumn },
        _react2.default.createElement(_assets2.default, { assets: this.props.assets })
      )
    );
  };

  return ElectrodeElectrifyReactComponent;
}(_react2.default.Component);

exports.default = ElectrodeElectrifyReactComponent;


ElectrodeElectrifyReactComponent.displayName = "ElectrodeElectrifyReactComponent";

ElectrodeElectrifyReactComponent.propTypes = {
  modules: _react.PropTypes.object,
  modulesByPkg: _react.PropTypes.object,
  totalSizeByPkg: _react.PropTypes.number,
  assets: _react.PropTypes.object
};

ElectrodeElectrifyReactComponent.defaultProps = {};
